File: lib/telegram.ts
import { appEnv, assertServerEnv } from "@/lib/env";

type CreateInviteLinkOptions = {
  expireSeconds?: number;
  memberLimit?: number;
  name?: string;
};

type TelegramApiResponse<T> = {
  ok: boolean;
  result?: T;
  description?: string;
  error_code?: number;
};

type ChatInviteLink = {
  invite_link: string;
  creator: { id: number; is_bot: boolean; first_name: string };
  creates_join_request: boolean;
  is_primary: boolean;
  is_revoked: boolean;
  name?: string;
  expire_date?: number;
  member_limit?: number;
};

type ChatMember = {
  status: "creator" | "administrator" | "member" | "restricted" | "left" | "kicked";
  user: { id: number };
};

async function callTelegramApi<T>(method: string, params: Record<string, unknown>): Promise<T> {
  const token = assertServerEnv("TELEGRAM_BOT_TOKEN", appEnv.telegramBotToken);
  const response = await fetch(`https://api.telegram.org/bot${token}/${method}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
    cache: "no-store"
  });

  const payload = (await response.json()) as TelegramApiResponse<T>;

  if (!response.ok || !payload.ok || payload.result === undefined) {
    throw new Error(
      `Telegram API ${method} failed: ${payload.error_code ?? response.status} ${payload.description ?? ""}`.trim()
    );
  }

  return payload.result;
}

export async function createVipInviteLink(options: CreateInviteLinkOptions = {}) {
  const chatId = assertServerEnv("TELEGRAM_GROUP_ID_VIP", appEnv.telegramGroupIdVip);
  const memberLimit = options.memberLimit ?? 1;
  const expireSeconds = options.expireSeconds ?? 60 * 60 * 24 * 7;
  const expireDate = Math.floor(Date.now() / 1000) + expireSeconds;

  const link = await callTelegramApi<ChatInviteLink>("createChatInviteLink", {
    chat_id: chatId,
    expire_date: expireDate,
    member_limit: memberLimit,
    name: options.name ?? `pro-${Date.now()}`
  });

  return {
    inviteLink: link.invite_link,
    expireDate: link.expire_date ?? expireDate,
    memberLimit: link.member_limit ?? memberLimit
  };
}

export async function revokeInviteLink(inviteLink: string) {
  const chatId = assertServerEnv("TELEGRAM_GROUP_ID_VIP", appEnv.telegramGroupIdVip);

  return callTelegramApi<ChatInviteLink>("revokeChatInviteLink", {
    chat_id: chatId,
    invite_link: inviteLink
  });
}

export async function kickFromVipGroup(telegramUserId: string | number) {
  const chatId = assertServerEnv("TELEGRAM_GROUP_ID_VIP", appEnv.telegramGroupIdVip);

  return callTelegramApi<boolean>("banChatMember", {
    chat_id: chatId,
    user_id: Number(telegramUserId),
    until_date: Math.floor(Date.now() / 1000) + 60
  });
}

export async function getVipMember(telegramUserId: string | number) {
  const chatId = assertServerEnv("TELEGRAM_GROUP_ID_VIP", appEnv.telegramGroupIdVip);

  return callTelegramApi<ChatMember>("getChatMember", {
    chat_id: chatId,
    user_id: Number(telegramUserId)
  });
}

export async function sendBotMessage(chatId: string | number, text: string) {
  return callTelegramApi<{ message_id: number }>("sendMessage", {
    chat_id: chatId,
    text,
    parse_mode: "HTML",
    disable_web_page_preview: true
  });
}

File: worker/app/telegram.py
from __future__ import annotations

import json
import logging
import os

import httpx

logger = logging.getLogger(__name__)


class TelegramConfigError(RuntimeError):
    pass


def _get_token() -> str:
    token = os.environ.get("TELEGRAM_BOT_TOKEN", "").strip()
    if not token:
        raise TelegramConfigError("TELEGRAM_BOT_TOKEN is not set")
    return token


def _get_public_channel() -> str:
    channel = (
        os.environ.get("TELEGRAM_CHANNEL_ID_PUBLIC", "").strip()
        or os.environ.get("TELEGRAM_CHANNEL_HANDLE", "").strip()
        or os.environ.get("TELEGRAM_CHANNEL_PUBLIC", "").strip()
    )
    if not channel:
        raise TelegramConfigError(
            "Set TELEGRAM_CHANNEL_ID_PUBLIC (numeric chat id) or TELEGRAM_CHANNEL_HANDLE"
        )
    return channel


async def send_channel_message(
    text: str,
    *,
    chat_id: str | None = None,
    parse_mode: str = "HTML",
    disable_web_page_preview: bool = True,
    reply_markup: dict | None = None,
) -> dict:
    token = _get_token()
    target = chat_id or _get_public_channel()

    url = f"https://api.telegram.org/bot{token}/sendMessage"
    payload: dict = {
        "chat_id": target,
        "text": text,
        "parse_mode": parse_mode,
        "disable_web_page_preview": disable_web_page_preview,
    }
    if reply_markup is not None:
        payload["reply_markup"] = json.dumps(reply_markup, ensure_ascii=False)

    async with httpx.AsyncClient(timeout=15.0) as client:
        response = await client.post(url, json=payload)
        data = response.json()

    if not data.get("ok"):
        raise RuntimeError(
            f"Telegram sendMessage failed: {data.get('error_code')} {data.get('description')}"
        )

    logger.info("telegram message sent to %s (id=%s)", target, data["result"]["message_id"])
    return data["result"]
